kernel/src/scan/mod.rs

```diff
@@ -92,25 +96,169 @@ impl ScanBuilder {
         let logical_schema = self
             .schema
             .unwrap_or_else(|| self.snapshot.schema().clone().into());
-        let (all_fields, read_fields, have_partition_cols) = get_state_info(
+        let state_info = get_state_info(
             logical_schema.as_ref(),
             &self.snapshot.metadata().partition_columns,
-            self.snapshot.column_mapping_mode,
         )?;
-        let physical_schema = Arc::new(StructType::new(read_fields));
 
-        // important! before a read/write to the table we must check it is supported
-        self.snapshot.protocol().ensure_read_supported()?;
+        // If we have a predicate, verify the columns it references and apply column mapping. First,
+        // get the set of references; use that to filter the schema to only the columns of interest
+        // (and verify that all referenced columns exist); then use the resulting logical/physical
+        // mappings to rewrite the expression with physical column names.
+        //
+        // NOTE: It is possible the predicate doesn't reference any schema; in that case it must
+        // consist entirely of literal expressions. We can't just evaluate it directly, because data
+        // skipping has specific rules about what kinds of expressions it supports. So just
+        // propagate the empty schema and let the skipping machinery deal with it.
+        let physical_predicate = match self.predicate {
+            Some(predicate) => Self::build_physical_predicate(predicate, &logical_schema)?,
+            None => PhysicalPredicate::None,
+        };
 
         Ok(Scan {
             snapshot: self.snapshot,
             logical_schema,
-            physical_schema,
-            predicate: self.predicate,
-            all_fields,
-            have_partition_cols,
+            physical_schema: Arc::new(StructType::new(state_info.read_fields)),
+            physical_predicate,
+            all_fields: state_info.all_fields,
+            have_partition_cols: state_info.have_partition_cols,
         })
     }
+
+    fn build_physical_predicate(
+        predicate: ExpressionRef,
+        logical_schema: &Schema,
+    ) -> DeltaResult<PhysicalPredicate> {
+        if can_statically_skip_all_files(&predicate) {
+            return Ok(PhysicalPredicate::StaticSkipAll);
+        }
+        let mut get_referenced_fields = GetReferencedFields {
+            references: predicate.references(),
+            column_mappings: HashMap::new(),
+            logical_path: vec![],
+            physical_path: vec![],
+        };
+        let schema_opt = get_referenced_fields.transform_struct(logical_schema);
+        if let Some(missing_column) = get_referenced_fields.references.into_iter().next() {
```

Comment:
would add a comment that explains that we expect `references` to be empty at this point because all were transformed into mappings